// @ant-design/pro-form/lib/components/List/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.FormListContext = void 0;

require("antd/es/form/style");

var _form = _interopRequireDefault(require("antd/es/form"));

require("antd/es/config-provider/style");

var _configProvider = _interopRequireDefault(require("antd/es/config-provider"));

require("antd/es/button/style");

var _button = _interopRequireDefault(require("antd/es/button"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

require("antd/es/tooltip/style");

var _tooltip = _interopRequireDefault(require("antd/es/tooltip"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _icons = require("@ant-design/icons");

var _proUtils = require("@ant-design/pro-utils");

var _omit = _interopRequireDefault(require("omit.js"));

var _toArray = _interopRequireDefault(require("rc-util/lib/Children/toArray"));

var _react = _interopRequireWildcard(require("react"));

require("./index.less");

var _warning = require("rc-util/lib/warning");

var _excluded = ["creatorButtonProps", "deleteIconProps", "copyIconProps", "itemContainerRender", "itemRender", "prefixCls", "creatorRecord", "action", "children", "actionRender", "fields", "meta", "field", "index", "formInstance", "alwaysShowItemLabel"],
    _excluded2 = ["actionRender", "creatorButtonProps", "label", "alwaysShowItemLabel", "tooltip", "creatorRecord", "itemRender", "rules", "itemContainerRender", "copyIconProps", "children", "deleteIconProps", "actionRef", "style", "prefixCls"];

var FormListContext = /*#__PURE__*/_react.default.createContext({});
/** Antd 自带的toArray 不这次方法，所以需要自己搞一个 */


exports.FormListContext = FormListContext;

var listToArray = function listToArray(children) {
  if (Array.isArray(children)) {
    return children;
  }

  if (typeof children === 'function') {
    return [children];
  }

  return (0, _toArray.default)(children);
};

var ProFormListItem = function ProFormListItem(props) {
  var _formInstance$getFiel;

  var creatorButtonProps = props.creatorButtonProps,
      deleteIconProps = props.deleteIconProps,
      copyIconProps = props.copyIconProps,
      itemContainerRender = props.itemContainerRender,
      itemRender = props.itemRender,
      prefixCls = props.prefixCls,
      creatorRecord = props.creatorRecord,
      action = props.action,
      children = props.children,
      actionRender = props.actionRender,
      fields = props.fields,
      meta = props.meta,
      field = props.field,
      index = props.index,
      formInstance = props.formInstance,
      alwaysShowItemLabel = props.alwaysShowItemLabel,
      rest = (0, _objectWithoutProperties2.default)(props, _excluded);
  var listContext = (0, _react.useContext)(FormListContext);
  var childrenArray = listToArray(children).map(function (childrenItem) {
    if (typeof childrenItem === 'function') {
      return childrenItem === null || childrenItem === void 0 ? void 0 : childrenItem(field, index, action);
    }

    return childrenItem;
  }).map(function (childrenItem) {
    if ( /*#__PURE__*/_react.default.isValidElement(childrenItem)) {
      return /*#__PURE__*/_react.default.cloneElement(childrenItem, (0, _objectSpread2.default)({
        key: childrenItem.key || (0, _proUtils.nanoid)()
      }, childrenItem === null || childrenItem === void 0 ? void 0 : childrenItem.props));
    }

    return childrenItem;
  });
  var copyIcon = (0, _react.useMemo)(function () {
    /** 复制按钮的配置 */
    if (!copyIconProps) return null;
    var _copyIconProps$Icon = copyIconProps.Icon,
        Icon = _copyIconProps$Icon === void 0 ? _icons.CopyOutlined : _copyIconProps$Icon,
        tooltipText = copyIconProps.tooltipText;
    return /*#__PURE__*/_react.default.createElement(_tooltip.default, {
      title: tooltipText,
      key: "copy"
    }, /*#__PURE__*/_react.default.createElement(Icon, {
      className: "".concat(prefixCls, "-action-icon"),
      onClick: function onClick() {
        action.add(formInstance === null || formInstance === void 0 ? void 0 : formInstance.getFieldValue([listContext.listName, rest.name, field.name].filter(function (item) {
          return item !== undefined;
        }).flat(1)));
      }
    }));
  }, [action, copyIconProps, field.name, formInstance, listContext.listName, prefixCls, rest.name]);
  var deleteIcon = (0, _react.useMemo)(function () {
    if (!deleteIconProps) return null;
    var _deleteIconProps$Icon = deleteIconProps.Icon,
        Icon = _deleteIconProps$Icon === void 0 ? _icons.DeleteOutlined : _deleteIconProps$Icon,
        tooltipText = deleteIconProps.tooltipText;
    return /*#__PURE__*/_react.default.createElement(_tooltip.default, {
      title: tooltipText,
      key: "delete"
    }, /*#__PURE__*/_react.default.createElement(Icon, {
      className: "".concat(prefixCls, "-action-icon"),
      onClick: function onClick() {
        return action.remove(field.name);
      }
    }));
  }, [action, deleteIconProps, field.name, prefixCls]);
  var defaultActionDom = (0, _react.useMemo)(function () {
    return [copyIcon, deleteIcon].filter(Boolean);
  }, [copyIcon, deleteIcon]);
  var actions = (actionRender === null || actionRender === void 0 ? void 0 : actionRender(field, action, defaultActionDom)) || defaultActionDom;
  var dom = actions.length > 0 ? /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefixCls, "-action")
  }, actions) : null;
  var options = {
    field: field,
    index: index,
    record: formInstance === null || formInstance === void 0 ? void 0 : (_formInstance$getFiel = formInstance.getFieldValue) === null || _formInstance$getFiel === void 0 ? void 0 : _formInstance$getFiel.call(formInstance, [listContext.listName, rest.name, field.name].filter(function (item) {
      return item !== undefined;
    }).flat(1)),
    fields: fields,
    operation: action,
    meta: meta
  };
  var itemContainer = (itemContainerRender === null || itemContainerRender === void 0 ? void 0 : itemContainerRender(childrenArray, options)) || childrenArray;

  var contentDom = (itemRender === null || itemRender === void 0 ? void 0 : itemRender({
    listDom: /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefixCls, "-container")
    }, itemContainer),
    action: dom
  }, options)) || /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefixCls, "-item").concat(alwaysShowItemLabel ? " ".concat(prefixCls, "-item-show-label") : ''),
    style: {
      display: 'flex',
      alignItems: 'flex-end'
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(prefixCls, "-container")
  }, itemContainer), dom);

  return /*#__PURE__*/_react.default.createElement(FormListContext.Provider, {
    key: field.name,
    value: (0, _objectSpread2.default)((0, _objectSpread2.default)({}, field), {}, {
      listName: [listContext.listName, rest.originName, field.name].filter(function (item) {
        return item !== undefined;
      }).flat(1)
    })
  }, contentDom);
};

var ProFormListContainer = function ProFormListContainer(props) {
  var creatorButtonProps = props.creatorButtonProps,
      prefixCls = props.prefixCls,
      children = props.children,
      creatorRecord = props.creatorRecord,
      action = props.action,
      fields = props.fields;
  var fieldKeyMap = (0, _react.useRef)(new Map());
  var uuidFields = (0, _react.useMemo)(function () {
    return fields.map(function (field) {
      var _fieldKeyMap$current, _fieldKeyMap$current3;

      if (!((_fieldKeyMap$current = fieldKeyMap.current) === null || _fieldKeyMap$current === void 0 ? void 0 : _fieldKeyMap$current.has(field.key.toString()))) {
        var _fieldKeyMap$current2;

        (_fieldKeyMap$current2 = fieldKeyMap.current) === null || _fieldKeyMap$current2 === void 0 ? void 0 : _fieldKeyMap$current2.set(field.key.toString(), (0, _proUtils.nanoid)());
      }

      var uuid = (_fieldKeyMap$current3 = fieldKeyMap.current) === null || _fieldKeyMap$current3 === void 0 ? void 0 : _fieldKeyMap$current3.get(field.key.toString());
      return (0, _objectSpread2.default)((0, _objectSpread2.default)({}, field), {}, {
        uuid: uuid
      });
    });
  }, [fields]);
  var creatorButton = (0, _react.useMemo)(function () {
    if (creatorButtonProps === false) return null;

    var _ref = creatorButtonProps || {},
        _ref$position = _ref.position,
        position = _ref$position === void 0 ? 'bottom' : _ref$position,
        _ref$creatorButtonTex = _ref.creatorButtonText,
        creatorButtonText = _ref$creatorButtonTex === void 0 ? '添加一行数据' : _ref$creatorButtonTex;

    return /*#__PURE__*/_react.default.createElement(_button.default, (0, _extends2.default)({
      className: "".concat(prefixCls, "-creator-button-").concat(position),
      type: "dashed",
      block: true,
      icon: /*#__PURE__*/_react.default.createElement(_icons.PlusOutlined, null)
    }, (0, _omit.default)(creatorButtonProps || {}, ['position', 'creatorButtonText']), {
      onClick: function onClick() {
        var index; // 如果是顶部，加到第一个，如果不是，为空就是最后一个

        if (position === 'top') index = 0;
        action.add((0, _proUtils.runFunction)(creatorRecord), index);
      }
    }), creatorButtonText);
  }, [action, creatorButtonProps, creatorRecord, prefixCls]);
  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width: 'max-content',
      maxWidth: '100%'
    }
  }, creatorButtonProps !== false && (creatorButtonProps === null || creatorButtonProps === void 0 ? void 0 : creatorButtonProps.position) === 'top' && creatorButton, uuidFields.map(function (field, index) {
    return /*#__PURE__*/_react.default.createElement(ProFormListItem, (0, _extends2.default)({}, props, {
      key: field.uuid,
      field: field,
      index: index
    }), children);
  }), creatorButtonProps !== false && (creatorButtonProps === null || creatorButtonProps === void 0 ? void 0 : creatorButtonProps.position) !== 'top' && creatorButton);
};

var ProFormList = function ProFormList(_ref2) {
  var actionRender = _ref2.actionRender,
      creatorButtonProps = _ref2.creatorButtonProps,
      label = _ref2.label,
      alwaysShowItemLabel = _ref2.alwaysShowItemLabel,
      tooltip = _ref2.tooltip,
      creatorRecord = _ref2.creatorRecord,
      itemRender = _ref2.itemRender,
      rules = _ref2.rules,
      itemContainerRender = _ref2.itemContainerRender,
      _ref2$copyIconProps = _ref2.copyIconProps,
      copyIconProps = _ref2$copyIconProps === void 0 ? {
    Icon: _icons.CopyOutlined,
    tooltipText: '复制此行'
  } : _ref2$copyIconProps,
      children = _ref2.children,
      _ref2$deleteIconProps = _ref2.deleteIconProps,
      deleteIconProps = _ref2$deleteIconProps === void 0 ? {
    Icon: _icons.DeleteOutlined,
    tooltipText: '删除此行'
  } : _ref2$deleteIconProps,
      actionRef = _ref2.actionRef,
      style = _ref2.style,
      prefixCls = _ref2.prefixCls,
      rest = (0, _objectWithoutProperties2.default)(_ref2, _excluded2);
  var actionRefs = (0, _react.useRef)();
  var context = (0, _react.useContext)(_configProvider.default.ConfigContext);
  var listContext = (0, _react.useContext)(FormListContext);
  var baseClassName = context.getPrefixCls('pro-form-list'); // 处理 list 的嵌套

  var name = (0, _react.useMemo)(function () {
    if (listContext.name === undefined) {
      return [rest.name].flat(1);
    }

    return [listContext.name, rest.name].flat(1);
  }, [listContext.name, rest.name]); // eslint-disable-next-line react-hooks/exhaustive-deps

  (0, _react.useImperativeHandle)(actionRef, function () {
    return actionRefs.current;
  }, [actionRefs.current]);
  var proFormContext = (0, _react.useContext)(_proUtils.ProFormContext);
  (0, _react.useEffect)(function () {
    (0, _warning.noteOnce)(!!proFormContext.formRef, "ProFormList \u5FC5\u987B\u8981\u653E\u5230 ProForm \u4E2D,\u5426\u5219\u4F1A\u9020\u6210\u884C\u4E3A\u5F02\u5E38\u3002");
    (0, _warning.noteOnce)(!!proFormContext.formRef, "Proformlist must be placed in ProForm, otherwise it will cause abnormal behavior.");
  }, [proFormContext.formRef]);
  if (!proFormContext.formRef) return null;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: baseClassName,
    style: style
  }, /*#__PURE__*/_react.default.createElement(_form.default.Item, (0, _extends2.default)({
    label: label,
    prefixCls: prefixCls,
    tooltip: tooltip,
    style: style
  }, rest, {
    name: undefined,
    rules: undefined
  }), /*#__PURE__*/_react.default.createElement(_form.default.List, (0, _extends2.default)({
    rules: rules
  }, rest, {
    name: name
  }), function (fields, action, meta) {
    // 将 action 暴露给外部
    actionRefs.current = action;
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(ProFormListContainer, {
      name: name,
      originName: rest.name,
      copyIconProps: copyIconProps,
      deleteIconProps: deleteIconProps,
      formInstance: proFormContext.formRef.current,
      prefixCls: baseClassName,
      meta: meta,
      fields: fields,
      itemContainerRender: itemContainerRender,
      itemRender: itemRender,
      creatorButtonProps: creatorButtonProps,
      creatorRecord: creatorRecord,
      actionRender: actionRender,
      action: action,
      alwaysShowItemLabel: alwaysShowItemLabel
    }, children), /*#__PURE__*/_react.default.createElement(_form.default.ErrorList, {
      errors: meta.errors
    }));
  })));
};

var _default = ProFormList;
exports.default = _default;
