// @ant-design/pro-field/lib/components/TimePicker/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.FieldTimeRangePicker = void 0;

require("antd/es/date-picker/style");

var _datePicker = _interopRequireDefault(require("antd/es/date-picker"));

require("antd/es/time-picker/style");

var _timePicker = _interopRequireDefault(require("antd/es/time-picker"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

require("antd/es/config-provider/style");

var _configProvider = _interopRequireDefault(require("antd/es/config-provider"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var _moment = _interopRequireDefault(require("moment"));

var _proUtils = require("@ant-design/pro-utils");

/**
 * 时间选择组件
 *
 * @param
 */
var FieldTimePicker = function FieldTimePicker(_ref, ref) {
  var text = _ref.text,
      mode = _ref.mode,
      light = _ref.light,
      label = _ref.label,
      format = _ref.format,
      render = _ref.render,
      renderFormItem = _ref.renderFormItem,
      plain = _ref.plain,
      fieldProps = _ref.fieldProps;

  var _useState = (0, _react.useState)(false),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      open = _useState2[0],
      setOpen = _useState2[1];

  var size = (0, _react.useContext)(_configProvider.default.SizeContext);

  var _useContext = (0, _react.useContext)(_configProvider.default.ConfigContext),
      getPrefixCls = _useContext.getPrefixCls;

  var prefixCls = getPrefixCls('pro-field-date-picker');
  var finalFormat = (fieldProps === null || fieldProps === void 0 ? void 0 : fieldProps.format) || format || 'HH:mm:ss';
  var isNumberOrMoment = _moment.default.isMoment(text) || typeof text === 'number';

  if (mode === 'read') {
    var dom = /*#__PURE__*/_react.default.createElement("span", {
      ref: ref
    }, text ? (0, _moment.default)(text, isNumberOrMoment ? undefined : finalFormat).format(finalFormat) : '-');

    if (render) {
      return render(text, (0, _objectSpread2.default)({
        mode: mode
      }, fieldProps), /*#__PURE__*/_react.default.createElement("span", null, dom));
    }

    return dom;
  }

  if (mode === 'edit' || mode === 'update') {
    var _dom;

    var disabled = fieldProps.disabled,
        _onChange = fieldProps.onChange,
        placeholder = fieldProps.placeholder,
        allowClear = fieldProps.allowClear,
        value = fieldProps.value;
    var momentValue = (0, _proUtils.parseValueToMoment)(value, finalFormat);

    if (light) {
      var valueStr = momentValue && momentValue.format(finalFormat) || '';
      _dom = /*#__PURE__*/_react.default.createElement("div", {
        className: "".concat(prefixCls, "-light"),
        onClick: function onClick() {
          setOpen(true);
        }
      }, /*#__PURE__*/_react.default.createElement(_timePicker.default, (0, _extends2.default)({
        value: momentValue,
        format: format,
        ref: ref
      }, fieldProps, {
        onChange: function onChange(v) {
          _onChange === null || _onChange === void 0 ? void 0 : _onChange(v);
          setTimeout(function () {
            setOpen(false);
          }, 0);
        },
        onOpenChange: setOpen,
        open: open
      })), /*#__PURE__*/_react.default.createElement(_proUtils.FieldLabel, {
        label: label,
        disabled: disabled,
        placeholder: placeholder,
        size: size,
        value: valueStr,
        allowClear: allowClear,
        onClear: function onClear() {
          _onChange === null || _onChange === void 0 ? void 0 : _onChange(null);
        },
        expanded: open
      }));
    } else {
      _dom = /*#__PURE__*/_react.default.createElement(_datePicker.default.TimePicker, (0, _extends2.default)({
        ref: ref,
        format: format,
        bordered: plain === undefined ? true : !plain
      }, fieldProps, {
        value: momentValue
      }));
    }

    if (renderFormItem) {
      return renderFormItem(text, (0, _objectSpread2.default)({
        mode: mode
      }, fieldProps), _dom);
    }

    return _dom;
  }

  return null;
};
/**
 * 时间区间选择
 *
 * @param param0
 * @param ref
 */


var FieldTimeRangePicker = function FieldTimeRangePicker(_ref2) {
  var text = _ref2.text,
      mode = _ref2.mode,
      format = _ref2.format,
      render = _ref2.render,
      renderFormItem = _ref2.renderFormItem,
      plain = _ref2.plain,
      fieldProps = _ref2.fieldProps;
  var finalFormat = (fieldProps === null || fieldProps === void 0 ? void 0 : fieldProps.format) || format || 'HH:mm:ss';

  var _ref3 = Array.isArray(text) ? text : [],
      _ref4 = (0, _slicedToArray2.default)(_ref3, 2),
      startText = _ref4[0],
      endText = _ref4[1];

  var startTextIsNumberOrMoment = _moment.default.isMoment(startText) || typeof startText === 'number';
  var endTextIsNumberOrMoment = _moment.default.isMoment(endText) || typeof endText === 'number';
  var parsedStartText = startText ? (0, _moment.default)(startText, startTextIsNumberOrMoment ? undefined : finalFormat).format(finalFormat) : '';
  var parsedEndText = endText ? (0, _moment.default)(endText, endTextIsNumberOrMoment ? undefined : finalFormat).format(finalFormat) : '';

  if (mode === 'read') {
    var dom = /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("div", null, parsedStartText || '-'), /*#__PURE__*/_react.default.createElement("div", null, parsedEndText || '-'));

    if (render) {
      return render(text, (0, _objectSpread2.default)({
        mode: mode
      }, fieldProps), /*#__PURE__*/_react.default.createElement("span", null, dom));
    }

    return dom;
  }

  if (mode === 'edit' || mode === 'update') {
    var value = fieldProps.value;
    var momentValue = (0, _proUtils.parseValueToMoment)(value, finalFormat);

    var _dom2 = /*#__PURE__*/_react.default.createElement(_timePicker.default.RangePicker, (0, _extends2.default)({
      format: format,
      bordered: plain === undefined ? true : !plain
    }, fieldProps, {
      value: momentValue
    }));

    if (renderFormItem) {
      return renderFormItem(text, (0, _objectSpread2.default)({
        mode: mode
      }, fieldProps), _dom2);
    }

    return _dom2;
  }

  return null;
};

exports.FieldTimeRangePicker = FieldTimeRangePicker;

var _default = /*#__PURE__*/_react.default.forwardRef(FieldTimePicker);

exports.default = _default;
