// @ant-design/pro-table/lib/components/Dropdown/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

require("antd/es/dropdown/style");

var _dropdown = _interopRequireDefault(require("antd/es/dropdown"));

require("antd/es/button/style");

var _button = _interopRequireDefault(require("antd/es/button"));

require("antd/es/menu/style");

var _menu = _interopRequireDefault(require("antd/es/menu"));

require("antd/es/config-provider/style");

var _configProvider = _interopRequireDefault(require("antd/es/config-provider"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _icons = require("@ant-design/icons");

require("./index.less");

var _excluded = ["key", "name"];

/**
 * 一个简单的下拉菜单
 *
 * @param param0
 */
var DropdownButton = function DropdownButton(_ref) {
  var children = _ref.children,
      menus = _ref.menus,
      onSelect = _ref.onSelect,
      className = _ref.className,
      style = _ref.style;

  var _useContext = (0, _react.useContext)(_configProvider.default.ConfigContext),
      getPrefixCls = _useContext.getPrefixCls;

  var tempClassName = getPrefixCls('pro-table-dropdown');

  var menu = /*#__PURE__*/_react.default.createElement(_menu.default, {
    onClick: function onClick(params) {
      return onSelect && onSelect(params.key);
    }
  }, menus === null || menus === void 0 ? void 0 : menus.map(function (item) {
    return /*#__PURE__*/_react.default.createElement(_menu.default.Item, {
      key: item.key
    }, item.name);
  }));

  return /*#__PURE__*/_react.default.createElement(_dropdown.default, {
    overlay: menu,
    className: (0, _classnames.default)(tempClassName, className)
  }, /*#__PURE__*/_react.default.createElement(_button.default, {
    style: style
  }, children, " ", /*#__PURE__*/_react.default.createElement(_icons.DownOutlined, null)));
};

var TableDropdown = function TableDropdown(_ref2) {
  var propsClassName = _ref2.className,
      style = _ref2.style,
      onSelect = _ref2.onSelect,
      _ref2$menus = _ref2.menus,
      menus = _ref2$menus === void 0 ? [] : _ref2$menus,
      children = _ref2.children;

  var _useContext2 = (0, _react.useContext)(_configProvider.default.ConfigContext),
      getPrefixCls = _useContext2.getPrefixCls;

  var className = getPrefixCls('pro-table-dropdown');

  var menu = /*#__PURE__*/_react.default.createElement(_menu.default, {
    onClick: function onClick(params) {
      return onSelect && onSelect(params.key);
    }
  }, menus.map(function (_ref3) {
    var key = _ref3.key,
        name = _ref3.name,
        rest = (0, _objectWithoutProperties2.default)(_ref3, _excluded);
    return /*#__PURE__*/_react.default.createElement(_menu.default.Item, (0, _extends2.default)({
      key: key
    }, rest), name);
  }));

  return /*#__PURE__*/_react.default.createElement(_dropdown.default, {
    overlay: menu,
    className: (0, _classnames.default)(className, propsClassName)
  }, /*#__PURE__*/_react.default.createElement("a", {
    style: style
  }, children || /*#__PURE__*/_react.default.createElement(_icons.EllipsisOutlined, null)));
};

TableDropdown.Button = DropdownButton;
var _default = TableDropdown;
exports.default = _default;
