// @ant-design/pro-form/es/layouts/LoginForm/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import "antd/es/config-provider/style";
import _ConfigProvider from "antd/es/config-provider";
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
var _excluded = ["logo", "message", "title", "subTitle", "actions", "children"];
import React, { useContext, useMemo } from 'react';
import ProForm from '../ProForm';
import { useIntl } from '@ant-design/pro-provider';
import './index.less';

function LoginForm(props) {
  var logo = props.logo,
      message = props.message,
      title = props.title,
      subTitle = props.subTitle,
      actions = props.actions,
      children = props.children,
      proFormProps = _objectWithoutProperties(props, _excluded);

  var intl = useIntl();

  var submitter = _objectSpread({
    searchConfig: {
      submitText: intl.getMessage('loginForm.submitText', '登录')
    },
    render: function render(_, dom) {
      return dom.pop();
    },
    submitButtonProps: {
      size: 'large',
      style: {
        width: '100%'
      }
    }
  }, proFormProps.submitter);

  var context = useContext(_ConfigProvider.ConfigContext);
  var baseClassName = context.getPrefixCls('pro-form-login');

  var getCls = function getCls(className) {
    return "".concat(baseClassName, "-").concat(className);
  };
  /** 生成logo 的dom，如果是string 设置为图片 如果是个 dom 就原样保留 */


  var logoDom = useMemo(function () {
    if (!logo) return null;

    if (typeof logo === 'string') {
      return /*#__PURE__*/React.createElement("img", {
        src: logo
      });
    }

    return logo;
  }, [logo]);
  return /*#__PURE__*/React.createElement("div", {
    className: getCls('container')
  }, /*#__PURE__*/React.createElement("div", {
    className: getCls('top')
  }, title || logoDom ? /*#__PURE__*/React.createElement("div", {
    className: getCls('header')
  }, logoDom ? /*#__PURE__*/React.createElement("span", {
    className: getCls('logo')
  }, logoDom) : null, title ? /*#__PURE__*/React.createElement("span", {
    className: getCls('title')
  }, title) : null) : null, subTitle ? /*#__PURE__*/React.createElement("div", {
    className: getCls('desc')
  }, subTitle) : null), /*#__PURE__*/React.createElement("div", {
    className: getCls('main')
  }, /*#__PURE__*/React.createElement(ProForm, _extends({
    isKeyPressSubmit: true,
    submitter: submitter
  }, proFormProps), message, children), actions ? /*#__PURE__*/React.createElement("div", {
    className: getCls('other')
  }, actions) : null));
}

export default LoginForm;
