// @ant-design/pro-form/lib/layouts/DrawerForm/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd/es/drawer/style");

var _drawer = _interopRequireDefault(require("antd/es/drawer"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

require("antd/es/config-provider/style");

var _configProvider = _interopRequireDefault(require("antd/es/config-provider"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireWildcard(require("react"));

var _useMergedState7 = _interopRequireDefault(require("rc-util/lib/hooks/useMergedState"));

var _reactDom = require("react-dom");

var _omit = _interopRequireDefault(require("omit.js"));

var _BaseForm = _interopRequireDefault(require("../../BaseForm"));

var _warning = require("rc-util/lib/warning");

var _scrollLocker = _interopRequireDefault(require("rc-util/lib/Dom/scrollLocker"));

var _excluded = ["children", "trigger", "onVisibleChange", "drawerProps", "onFinish", "title", "width"];

function DrawerForm(_ref) {
  var _rest$submitter, _rest$submitter2;

  var children = _ref.children,
      trigger = _ref.trigger,
      onVisibleChange = _ref.onVisibleChange,
      drawerProps = _ref.drawerProps,
      onFinish = _ref.onFinish,
      title = _ref.title,
      width = _ref.width,
      rest = (0, _objectWithoutProperties2.default)(_ref, _excluded);

  var _useMergedState = (0, _useMergedState7.default)(!!rest.visible, {
    value: rest.visible,
    onChange: onVisibleChange
  }),
      _useMergedState2 = (0, _slicedToArray2.default)(_useMergedState, 2),
      visible = _useMergedState2[0],
      setVisible = _useMergedState2[1];

  var _useMergedState3 = (0, _useMergedState7.default)(0),
      _useMergedState4 = (0, _slicedToArray2.default)(_useMergedState3, 2),
      key = _useMergedState4[0],
      setKey = _useMergedState4[1];

  var context = (0, _react.useContext)(_configProvider.default.ConfigContext);
  /** Modal dom 解除渲染之后 */

  var _useMergedState5 = (0, _useMergedState7.default)(!!rest.visible),
      _useMergedState6 = (0, _slicedToArray2.default)(_useMergedState5, 2),
      isDestroy = _useMergedState6[0],
      setIsDestroy = _useMergedState6[1];

  (0, _react.useEffect)(function () {
    setIsDestroy(!!rest.visible); // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [rest.visible]);
  var renderDom = (0, _react.useMemo)(function () {
    var _context$getPopupCont;

    if ((drawerProps === null || drawerProps === void 0 ? void 0 : drawerProps.getContainer) === false) {
      return false;
    }

    if (drawerProps === null || drawerProps === void 0 ? void 0 : drawerProps.getContainer) {
      if (typeof (drawerProps === null || drawerProps === void 0 ? void 0 : drawerProps.getContainer) === 'function') {
        var _drawerProps$getConta;

        return drawerProps === null || drawerProps === void 0 ? void 0 : (_drawerProps$getConta = drawerProps.getContainer) === null || _drawerProps$getConta === void 0 ? void 0 : _drawerProps$getConta.call(drawerProps);
      }

      if (typeof (drawerProps === null || drawerProps === void 0 ? void 0 : drawerProps.getContainer) === 'string') {
        return document.getElementById(drawerProps === null || drawerProps === void 0 ? void 0 : drawerProps.getContainer);
      }

      return drawerProps === null || drawerProps === void 0 ? void 0 : drawerProps.getContainer;
    }

    return context === null || context === void 0 ? void 0 : (_context$getPopupCont = context.getPopupContainer) === null || _context$getPopupCont === void 0 ? void 0 : _context$getPopupCont.call(context, document.body); // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [context, drawerProps, visible]);

  var _useState = (0, _react.useState)(function () {
    if (typeof window === 'undefined') return undefined;
    return new _scrollLocker.default({
      container: renderDom || document.body
    });
  }),
      _useState2 = (0, _slicedToArray2.default)(_useState, 1),
      scrollLocker = _useState2[0];

  (0, _warning.noteOnce)( // eslint-disable-next-line @typescript-eslint/dot-notation
  !rest['footer'] || !(drawerProps === null || drawerProps === void 0 ? void 0 : drawerProps.footer), 'DrawerForm 是一个 ProForm 的特殊布局，如果想自定义按钮，请使用 submit.render 自定义。');
  (0, _react.useEffect)(function () {
    if (visible) {
      scrollLocker === null || scrollLocker === void 0 ? void 0 : scrollLocker.lock();
    } else {
      scrollLocker === null || scrollLocker === void 0 ? void 0 : scrollLocker.unLock();
    }

    if (visible && rest.visible) {
      onVisibleChange === null || onVisibleChange === void 0 ? void 0 : onVisibleChange(true);
    } // 如果打开了窗口，并且是用户设置的 visible，就触发一下重新更新


    if (visible && rest.visible && (drawerProps === null || drawerProps === void 0 ? void 0 : drawerProps.destroyOnClose)) {
      setKey(key + 1);
    }

    return function () {
      var _scrollLocker$unLock;

      if (!visible) scrollLocker === null || scrollLocker === void 0 ? void 0 : (_scrollLocker$unLock = scrollLocker.unLock) === null || _scrollLocker$unLock === void 0 ? void 0 : _scrollLocker$unLock.call(scrollLocker);
    }; // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [visible]);
  /** 设置 trigger 的情况下，懒渲染优化性能；使之可以直接配合表格操作等场景使用 */

  var isFirstRender = (0, _react.useRef)(!(drawerProps === null || drawerProps === void 0 ? void 0 : drawerProps.forceRender));
  /**
   * IsFirstRender.current 或者 visible 为 true 的时候就渲染 不渲染能会造成一些问题,比如再次打开值不对了 只有手动配置
   * drawerProps?.destroyOnClose 为 true 的时候才会每次关闭的时候删除 dom
   */

  var shouldRenderFormItems = (0, _react.useMemo)(function () {
    if (isFirstRender.current && visible === false) {
      return false;
    }

    if (visible === false && (drawerProps === null || drawerProps === void 0 ? void 0 : drawerProps.destroyOnClose)) {
      return false;
    }

    return true;
  }, [visible, drawerProps === null || drawerProps === void 0 ? void 0 : drawerProps.destroyOnClose]);
  /** 同步 props 和 本地 */

  var formRef = (0, _react.useRef)();
  /** 如果 destroyOnClose ，重置一下表单 */

  (0, _react.useEffect)(function () {
    if (visible) {
      isFirstRender.current = false;
    }
  }, [drawerProps === null || drawerProps === void 0 ? void 0 : drawerProps.destroyOnClose, visible]);
  (0, _react.useEffect)(function () {
    return function () {
      var _scrollLocker$unLock2;

      scrollLocker === null || scrollLocker === void 0 ? void 0 : (_scrollLocker$unLock2 = scrollLocker.unLock) === null || _scrollLocker$unLock2 === void 0 ? void 0 : _scrollLocker$unLock2.call(scrollLocker);
    };
  }, // eslint-disable-next-line react-hooks/exhaustive-deps
  []);
  (0, _react.useImperativeHandle)(rest.formRef, function () {
    return formRef.current;
  });

  var formDom = /*#__PURE__*/_react.default.createElement("div", {
    onClick: function onClick(e) {
      return e.stopPropagation();
    }
  }, /*#__PURE__*/_react.default.createElement(_BaseForm.default, (0, _extends2.default)({
    formComponentType: "DrawerForm",
    layout: "vertical",
    key: key
  }, (0, _omit.default)(rest, ['visible']), {
    formRef: formRef,
    onInit: function onInit(value, form) {
      var _rest$onInit;

      if (rest.formRef) {
        rest.formRef.current = form;
      }

      rest === null || rest === void 0 ? void 0 : (_rest$onInit = rest.onInit) === null || _rest$onInit === void 0 ? void 0 : _rest$onInit.call(rest, value, form);
    },
    submitter: rest.submitter === false ? false : (0, _objectSpread2.default)((0, _objectSpread2.default)({}, rest.submitter), {}, {
      searchConfig: (0, _objectSpread2.default)({
        submitText: '确认',
        resetText: '取消'
      }, (_rest$submitter = rest.submitter) === null || _rest$submitter === void 0 ? void 0 : _rest$submitter.searchConfig),
      resetButtonProps: (0, _objectSpread2.default)({
        preventDefault: true,
        onClick: function onClick(e) {
          var _drawerProps$onClose;

          setVisible(false);
          drawerProps === null || drawerProps === void 0 ? void 0 : (_drawerProps$onClose = drawerProps.onClose) === null || _drawerProps$onClose === void 0 ? void 0 : _drawerProps$onClose.call(drawerProps, e);
        }
      }, (_rest$submitter2 = rest.submitter) === null || _rest$submitter2 === void 0 ? void 0 : _rest$submitter2.resetButtonProps)
    }),
    onFinish: /*#__PURE__*/function () {
      var _ref2 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee(values) {
        var success;
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (onFinish) {
                  _context.next = 2;
                  break;
                }

                return _context.abrupt("return");

              case 2:
                _context.next = 4;
                return onFinish(values);

              case 4:
                success = _context.sent;

                if (success) {
                  setVisible(false);
                  setTimeout(function () {
                    var _formRef$current;

                    if (drawerProps === null || drawerProps === void 0 ? void 0 : drawerProps.destroyOnClose) (_formRef$current = formRef.current) === null || _formRef$current === void 0 ? void 0 : _formRef$current.resetFields();
                  }, 300);
                }

              case 6:
              case "end":
                return _context.stop();
            }
          }
        }, _callee);
      }));

      return function (_x) {
        return _ref2.apply(this, arguments);
      };
    }(),
    contentRender: function contentRender(item, submitter) {
      return /*#__PURE__*/_react.default.createElement(_drawer.default, (0, _extends2.default)({
        title: title,
        width: width || 800
      }, drawerProps, {
        getContainer: false,
        visible: visible,
        onClose: function onClose(e) {
          var _drawerProps$onClose2;

          setVisible(false);
          drawerProps === null || drawerProps === void 0 ? void 0 : (_drawerProps$onClose2 = drawerProps.onClose) === null || _drawerProps$onClose2 === void 0 ? void 0 : _drawerProps$onClose2.call(drawerProps, e); // drawer 的after close 在暂时有点问题，先用这个顶一下

          setTimeout(function () {
            setIsDestroy(false);
          }, 300);
        },
        footer: !!submitter && /*#__PURE__*/_react.default.createElement("div", {
          style: {
            display: 'flex',
            justifyContent: 'flex-end'
          }
        }, submitter)
      }), shouldRenderFormItems ? item : null);
    }
  }), children));
  /** 这个是为了支持 ssr */


  var portalRenderDom = (0, _react.useMemo)(function () {
    if (typeof window === 'undefined') return undefined;
    return renderDom || document.body;
  }, [renderDom]);

  var triggerDom = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, {
    key: "trigger"
  }, trigger && /*#__PURE__*/_react.default.cloneElement(trigger, (0, _objectSpread2.default)((0, _objectSpread2.default)({}, trigger.props), {}, {
    onClick: function () {
      var _onClick = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee2(e) {
        var _trigger$props, _trigger$props$onClic;

        return _regenerator.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                setVisible(!visible);
                setIsDestroy(!visible);
                (_trigger$props = trigger.props) === null || _trigger$props === void 0 ? void 0 : (_trigger$props$onClic = _trigger$props.onClick) === null || _trigger$props$onClic === void 0 ? void 0 : _trigger$props$onClic.call(_trigger$props, e);

              case 3:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2);
      }));

      function onClick(_x2) {
        return _onClick.apply(this, arguments);
      }

      return onClick;
    }()
  })));
  /** 如果destroyOnClose，关闭的时候解除渲染Form */


  if ((drawerProps === null || drawerProps === void 0 ? void 0 : drawerProps.destroyOnClose) && !isDestroy) return triggerDom;
  /** 不放到 body 上会导致 z-index 的问题 遮罩什么的都遮不住了 */

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, renderDom !== false && portalRenderDom ? /*#__PURE__*/(0, _reactDom.createPortal)(formDom, portalRenderDom) : formDom, triggerDom);
}

var _default = DrawerForm;
exports.default = _default;
