// @ant-design/pro-form/lib/components/Field/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));

var _react = _interopRequireDefault(require("react"));

var _proField = _interopRequireDefault(require("@ant-design/pro-field"));

var _proUtils = require("@ant-design/pro-utils");

var _createField = _interopRequireDefault(require("../../BaseForm/createField"));

var _Dependency = _interopRequireDefault(require("../Dependency"));

var _excluded = ["fieldProps", "children", "labelCol", "label", "autoFocus", "isDefaultDom", "render", "proFieldProps", "renderFormItem", "valueType", "initialValue", "onChange", "valueEnum", "params", "name", "cacheForSwr", "valuePropName"],
    _excluded2 = ["fieldProps", "children", "labelCol", "label", "autoFocus", "isDefaultDom", "render", "proFieldProps", "renderFormItem", "valueType", "initialValue", "onChange", "valueEnum", "params", "name"];

var ProFormField = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(ProFormField, _React$Component);

  var _super = (0, _createSuper2.default)(ProFormField);

  function ProFormField() {
    var _this;

    (0, _classCallCheck2.default)(this, ProFormField);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    _this.filedRef = /*#__PURE__*/_react.default.createRef();
    return _this;
  }

  (0, _createClass2.default)(ProFormField, [{
    key: "renderDom",
    value: function renderDom(values) {
      var _this$props = this.props,
          fieldProps = _this$props.fieldProps,
          children = _this$props.children,
          labelCol = _this$props.labelCol,
          label = _this$props.label,
          autoFocus = _this$props.autoFocus,
          isDefaultDom = _this$props.isDefaultDom,
          render = _this$props.render,
          proFieldProps = _this$props.proFieldProps,
          renderFormItem = _this$props.renderFormItem,
          valueType = _this$props.valueType,
          initialValue = _this$props.initialValue,
          _onChange = _this$props.onChange,
          valueEnum = _this$props.valueEnum,
          params = _this$props.params,
          name = _this$props.name,
          _this$props$cacheForS = _this$props.cacheForSwr,
          cacheForSwr = _this$props$cacheForS === void 0 ? false : _this$props$cacheForS,
          _this$props$valueProp = _this$props.valuePropName,
          valuePropName = _this$props$valueProp === void 0 ? 'value' : _this$props$valueProp,
          restProps = (0, _objectWithoutProperties2.default)(_this$props, _excluded);
      var propsParams = values ? (0, _objectSpread2.default)((0, _objectSpread2.default)({}, params), values || {}) : params;
      return /*#__PURE__*/_react.default.createElement(_proField.default, (0, _extends2.default)({
        valuePropName: valuePropName,
        text: fieldProps === null || fieldProps === void 0 ? void 0 : fieldProps[valuePropName],
        render: render,
        renderFormItem: renderFormItem,
        valueType: valueType || 'text',
        cacheForSwr: cacheForSwr,
        fieldProps: (0, _objectSpread2.default)((0, _objectSpread2.default)({
          autoFocus: autoFocus
        }, fieldProps), {}, {
          onChange: function onChange() {
            for (var _len2 = arguments.length, restParams = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
              restParams[_key2] = arguments[_key2];
            }

            if (fieldProps === null || fieldProps === void 0 ? void 0 : fieldProps.onChange) {
              var _fieldProps$onChange;

              fieldProps === null || fieldProps === void 0 ? void 0 : (_fieldProps$onChange = fieldProps.onChange) === null || _fieldProps$onChange === void 0 ? void 0 : _fieldProps$onChange.call.apply(_fieldProps$onChange, [fieldProps].concat(restParams));
              return;
            }

            _onChange === null || _onChange === void 0 ? void 0 : _onChange.apply(void 0, restParams);
          }
        }),
        valueEnum: (0, _proUtils.runFunction)(valueEnum)
      }, proFieldProps, restProps, {
        mode: (proFieldProps === null || proFieldProps === void 0 ? void 0 : proFieldProps.mode) || 'edit',
        params: propsParams,
        ref: this.filedRef
      }));
    }
  }, {
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps) {
      return !(0, _proUtils.isDeepEqualReact)(nextProps, this.props, ['onChange', 'onBlur']);
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props2 = this.props,
          fieldProps = _this$props2.fieldProps,
          children = _this$props2.children,
          labelCol = _this$props2.labelCol,
          label = _this$props2.label,
          autoFocus = _this$props2.autoFocus,
          isDefaultDom = _this$props2.isDefaultDom,
          render = _this$props2.render,
          proFieldProps = _this$props2.proFieldProps,
          renderFormItem = _this$props2.renderFormItem,
          valueType = _this$props2.valueType,
          initialValue = _this$props2.initialValue,
          _onChange2 = _this$props2.onChange,
          valueEnum = _this$props2.valueEnum,
          params = _this$props2.params,
          name = _this$props2.name,
          restProps = (0, _objectWithoutProperties2.default)(_this$props2, _excluded2); // 防止 formItem 的值被吃掉

      if (children) {
        if ( /*#__PURE__*/_react.default.isValidElement(children)) {
          return /*#__PURE__*/_react.default.cloneElement(children, (0, _objectSpread2.default)((0, _objectSpread2.default)({}, restProps), {}, {
            onChange: function onChange() {
              for (var _len3 = arguments.length, restParams = new Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
                restParams[_key3] = arguments[_key3];
              }

              if (fieldProps === null || fieldProps === void 0 ? void 0 : fieldProps.onChange) {
                var _fieldProps$onChange2;

                fieldProps === null || fieldProps === void 0 ? void 0 : (_fieldProps$onChange2 = fieldProps.onChange) === null || _fieldProps$onChange2 === void 0 ? void 0 : _fieldProps$onChange2.call.apply(_fieldProps$onChange2, [fieldProps].concat(restParams));
                return;
              }

              _onChange2 === null || _onChange2 === void 0 ? void 0 : _onChange2.apply(void 0, restParams);
            }
          }, children.props));
        }

        return children;
      }

      if (restProps.dependencies && restProps.request) {
        return /*#__PURE__*/_react.default.createElement(_Dependency.default, {
          name: restProps.dependencies
        }, function (values) {
          return _this2.renderDom(values);
        });
      }

      return this.renderDom();
    }
  }]);
  return ProFormField;
}(_react.default.Component);

var _default = (0, _createField.default)(ProFormField);

exports.default = _default;
