// @ant-design/pro-field/es/components/TreeSelect/index.js
import "antd/es/spin/style";
import _Spin from "antd/es/spin";
import "antd/es/tree-select/style";
import _TreeSelect from "antd/es/tree-select";
import _extends from "@babel/runtime/helpers/esm/extends";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import "antd/es/config-provider/style";
import _ConfigProvider from "antd/es/config-provider";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
var _excluded = ["radioType", "renderFormItem", "mode", "render"],
    _excluded2 = ["onSearch", "onClear", "onChange", "onBlur", "loadData", "showSearch", "autoClearSearchValue", "searchValue"];
import React, { useContext, useMemo, useImperativeHandle, useRef, useState } from 'react';
import classNames from 'classnames';
import { ObjToMap, proFieldParsingText, useFieldFetchData } from '../Select';
/**
 * Tree select
 *
 * @param param0
 * @param ref
 */

var FieldTreeSelect = function FieldTreeSelect(_ref, ref) {
  var radioType = _ref.radioType,
      renderFormItem = _ref.renderFormItem,
      mode = _ref.mode,
      render = _ref.render,
      rest = _objectWithoutProperties(_ref, _excluded);

  var _useContext = useContext(_ConfigProvider.ConfigContext),
      getPrefixCls = _useContext.getPrefixCls;

  var layoutClassName = getPrefixCls('pro-field-tree-select');
  var treeSelectRef = useRef(null);

  var _ref2 = rest.fieldProps || {},
      _onSearch = _ref2.onSearch,
      _onClear = _ref2.onClear,
      propsOnChange = _ref2.onChange,
      _onBlur = _ref2.onBlur,
      loadData = _ref2.loadData,
      showSearch = _ref2.showSearch,
      autoClearSearchValue = _ref2.autoClearSearchValue,
      _ref2$searchValue = _ref2.searchValue,
      propsSearchValue = _ref2$searchValue === void 0 ? '' : _ref2$searchValue,
      fieldProps = _objectWithoutProperties(_ref2, _excluded2);

  var _useFieldFetchData = useFieldFetchData(_objectSpread(_objectSpread({}, rest), {}, {
    defaultKeyWords: propsSearchValue
  })),
      _useFieldFetchData2 = _slicedToArray(_useFieldFetchData, 3),
      loading = _useFieldFetchData2[0],
      options = _useFieldFetchData2[1],
      _fetchData = _useFieldFetchData2[2];

  var _useState = useState(propsSearchValue),
      _useState2 = _slicedToArray(_useState, 2),
      searchValue = _useState2[0],
      setSearchValue = _useState2[1];

  useImperativeHandle(ref, function () {
    return _objectSpread(_objectSpread({}, treeSelectRef.current || {}), {}, {
      fetchData: function fetchData() {
        return _fetchData();
      }
    });
  });
  var optionsValueEnum = useMemo(function () {
    /**
     * Support TreeSelect fieldNames
     *
     * @see https://ant.design/components/tree-select-cn
     */
    var _ref3 = (fieldProps === null || fieldProps === void 0 ? void 0 : fieldProps.fieldNames) || {},
        _ref3$value = _ref3.value,
        valuePropsName = _ref3$value === void 0 ? 'value' : _ref3$value,
        _ref3$label = _ref3.label,
        labelPropsName = _ref3$label === void 0 ? 'label' : _ref3$label,
        _ref3$children = _ref3.children,
        childrenPropsName = _ref3$children === void 0 ? 'children' : _ref3$children;

    var traverseOptions = function traverseOptions(_options) {
      return (_options === null || _options === void 0 ? void 0 : _options.length) > 0 ? _options === null || _options === void 0 ? void 0 : _options.reduce(function (pre, cur) {
        var label = cur[labelPropsName],
            value = cur[valuePropsName],
            children = cur[childrenPropsName];
        return _objectSpread(_objectSpread({}, pre), {}, _defineProperty({}, value, label), traverseOptions(children));
      }, {}) : {};
    };

    return traverseOptions(options);
  }, [options, fieldProps === null || fieldProps === void 0 ? void 0 : fieldProps.fieldNames]);

  var onChange = function onChange(value, optionList, extra) {
    // 将搜索框置空 和 antd 行为保持一致
    if (showSearch && autoClearSearchValue) {
      _fetchData('');

      _onSearch === null || _onSearch === void 0 ? void 0 : _onSearch('');
      setSearchValue('');
    }
    /** Fix: TreeSelect warningProps 和TreeSelect结果保持一致 */


    propsOnChange === null || propsOnChange === void 0 ? void 0 : propsOnChange(value, optionList, extra);
  };

  if (mode === 'read') {
    var dom = /*#__PURE__*/React.createElement(React.Fragment, null, proFieldParsingText(rest.text, ObjToMap(rest.valueEnum || optionsValueEnum)));

    if (render) {
      return render(rest.text, _objectSpread({
        mode: mode
      }, fieldProps), dom) || null;
    }

    return dom;
  }

  if (mode === 'edit') {
    var _dom = /*#__PURE__*/React.createElement(_Spin, {
      spinning: loading
    }, /*#__PURE__*/React.createElement(_TreeSelect, _extends({
      ref: treeSelectRef
    }, fieldProps, {
      treeData: options,
      showSearch: showSearch,
      searchValue: searchValue,
      autoClearSearchValue: autoClearSearchValue,
      onClear: function onClear() {
        _onClear === null || _onClear === void 0 ? void 0 : _onClear();

        _fetchData('');

        if (showSearch) {
          setSearchValue('');
        }
      },
      loadData: loadData ? function (node) {
        var _node$value;

        _fetchData((_node$value = node.value) === null || _node$value === void 0 ? void 0 : _node$value.toString());

        return (loadData === null || loadData === void 0 ? void 0 : loadData(node)) || Promise.resolve();
      } : undefined,
      onChange: onChange,
      onSearch: function onSearch(value) {
        _fetchData(value);

        _onSearch === null || _onSearch === void 0 ? void 0 : _onSearch(value);
        setSearchValue(value);
      },
      onBlur: function onBlur(event) {
        setSearchValue('');

        _fetchData('');

        _onBlur === null || _onBlur === void 0 ? void 0 : _onBlur(event);
      },
      className: classNames(fieldProps === null || fieldProps === void 0 ? void 0 : fieldProps.className, layoutClassName)
    })));

    if (renderFormItem) {
      return renderFormItem(rest.text, _objectSpread({
        mode: mode
      }, fieldProps), _dom) || null;
    }

    return _dom;
  }

  return null;
};

export default /*#__PURE__*/React.forwardRef(FieldTreeSelect);
